### PR TITLE
New Rule: param-property-in-constructor

### PR DESCRIPTION
**Pull request by chrismbarr**
New rule created based on discussions in #1336

The rule is named `no-improper-constructor-param-usage` and has no configuration options.

Here’s a quick code sample of what will trigger a violation
```ts
constructor(public foo: number) {
    this.foo = 10;
    foo = 5; //<-- violation
}
```

### CODE REVIEW

**Comment by chrismbarr** [comment]
Sorry about the merge commits.  I thought github desktop was better than that!  I'll do it from the command line in the future.

**Comment by JKillian** [comment]
How do you feel about the name `param-property-in-constructor`?  "no-improper" feels like an uneeded double-negative to me, and "usage" seems unnecessary. "Parameter properties" is the official name the TS docs use for this sort of a construct, so I wanted to get that in the rule name

File: src/rules/noImproperConstructorParamUsageRule.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        const languageService = Lint.createLanguageService(sourceFile.fileName, sourceFile.getFullText());
+        return this.applyWithWalker(new NoImproperCtorParamUsageWalker(sourceFile, this.getOptions(), languageService));
+    }
+}
+
+export class NoImproperCtorParamUsageWalker extends Lint.RuleWalker {
+
+    private languageService: ts.LanguageService;
+
+    constructor(sourceFile: ts.SourceFile, options: Lint.IOptions, languageService: ts.LanguageService) {
```
Comment (JKillian) [inline src/rules/noImproperConstructorParamUsageRule.ts]: You could just make this `private languageService: ts.LanguageService` if you want (funny considering what this rule does 😄 )

File: test/rules/no-improper-constructor-param-usage/test.ts.lint [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        this.num = 10;
+        num = 5;
+        ~~~ [The constructor parameter 'num' should only be accessed with the 'this' keyword: 'this.num']
+
+        animal = "aardvark";
+        ~~~~~~ [The constructor parameter 'animal' should only be accessed with the 'this' keyword: 'this.animal']
+        this.animal = "skunk";
+
+        fruit = "banana";
+        ~~~~~ [The constructor parameter 'fruit' should only be accessed with the 'this' keyword: 'this.fruit']
```
Comment (JKillian) [inline test/rules/no-improper-constructor-param-usage/test.ts.lint]: This seems like a bug to me, `fruit` is just a regular old parameter, so this should be allowed, right? (Also occurs below)

File: src/rules/noImproperConstructorParamUsageRule.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            const fileName = this.getSourceFile().fileName;
+            node.parameters.forEach((param: ts.ParameterDeclaration) => {
+                this.validateParam(param, fileName);
+            });
+        }
+        super.visitConstructorDeclaration(node);
+    }
+
+    private validateParam(param: ts.ParameterDeclaration, fileName: string) {
+        const highlights = this.languageService.getDocumentHighlights(fileName, param.name.getStart(), [fileName]);
```
Comment (JKillian) [inline src/rules/noImproperConstructorParamUsageRule.ts]: We only need to validate if the parameter is actually a parameter property, right? We check for that in another rule like this: https://github.com/palantir/tslint/blob/81c538c2cdc03bf319f5953bd573271420ff978a/src/rules/noConstructorVarsRule.ts#L48

File: src/rules/noImproperConstructorParamUsageRule.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    constructor(sourceFile: ts.SourceFile, options: Lint.IOptions, languageService: ts.LanguageService) {
+        super(sourceFile, options);
+        this.languageService = languageService;
+    }
+
+    public visitConstructorDeclaration(node: ts.ConstructorDeclaration) {
+        if (node.parameters && node.parameters.length > 0) {
+            const fileName = this.getSourceFile().fileName;
+            node.parameters.forEach((param: ts.ParameterDeclaration) => {
```
Comment (JKillian) [inline src/rules/noImproperConstructorParamUsageRule.ts]: `for-of` usage is preferred when you don't need an index argument

Comment (adidahiya) [inline src/rules/noImproperConstructorParamUsageRule.ts]: Performance and consistency (one iteration style)

File: src/rules/noImproperConstructorParamUsageRule.ts (hunk 1 above)
Comment (chrismbarr) [inline src/rules/noImproperConstructorParamUsageRule.ts]: lol, ugh why did I not notice this! :laughing:

File: test/rules/no-improper-constructor-param-usage/test.ts.lint (hunk 2 above)
Comment (chrismbarr) [inline test/rules/no-improper-constructor-param-usage/test.ts.lint]: Ah good point.  i suppose this rule should only check params that have a `public` or `private` modifier on it, correct?

**Comment by chrismbarr** [comment]
The test failures seem to be related to changes that were made before this branch was started.  I think it would be fine to merge into master, but please double check me.

**Comment by JKillian** [comment]
Can you merge master here as well, so we can see the CI passing before merging this PR?

File: test/rules/param-property-in-constructor/test.ts.lint [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+        mineral = "Gypsum";
+        ~~~~~~~ [The constructor parameter 'mineral' should only be accessed with the 'this' keyword: 'this.mineral']
+        this.mineral = "Fluorite";
+    }
+}
+
+class FailingExample2 {
+    constructor(public thing: Object) {
+        const wow=thing.doTheThing()
```
Comment (JKillian) [inline test/rules/param-property-in-constructor/test.ts.lint]: Hmm, I wasn't thinking about this use case. When accessing a value, it doesn't matter if you use `this.` or not. Do you think we should still disallow this?

Comment (JKillian) [inline test/rules/param-property-in-constructor/test.ts.lint]: On the other hand, using `this.` doesn't hurt

Comment (chrismbarr) [inline test/rules/param-property-in-constructor/test.ts.lint]: yeah that was my thinking with the way it was written initially.  I was thinking that if we just require `this.` on everything, then there is consistency for how all constructor params are used, regardless if they are used as properties or not.

**Comment by JKillian** [comment]
Hmm, actually looks like TS has some sort of similar functionality build in: https://github.com/Microsoft/TypeScript/pull/10062

**Comment by YuichiNukiyama** [comment]
@JKillian You're right. We can validate unused parameter property  with `--noUnusedParameters ` option.
This feature already has been implimented in TS2.0 beta. And my PR (#10062) only intend to change error message. see [Flag unused declarations with --noUnusedParameters and --noUnusedLocals ](https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#flag-unused-declarations-with---nounusedparameters-and---nounusedlocals).

**Comment by adidahiya** [comment]
related: #1481

**Comment by JKillian** [comment]
I don't think the benefits of this rule are worth it given the features in TS 2.0. Sorry @chrismbarr! Let me know if you strongly disagree

**Comment by chrismbarr** [comment]
@JKillian :Some code I saw today in a project lead me back here to see why this PR was never accepted.  Reading over the above comments, I'm not entirely sure now.

The intent of this PR was **not** to find unused constructor parameters, it was meant to **enforce the correct usage** of those parameters.

```ts
constructor(public foo: number) {
    this.foo = 10;
    foo = 5; //<-- violation
}
```

I can re-submit this PR, but I don't want to put the work in for it unless it's desired.

**Pull request closed without merging**